### ISSUE TITLE
Differences in how vignette's build between devtools::build_vignettes and devtools::check() in R 4.0.2 which doesn't occur in R 3.6.0

### ISSUE THREAD

**Issue by NathanSkene** [r-lib/devtools#2262]
Hi,

I'm trying to get devtools::check() to complete successfully for my package but have hit a somewhat intractable issue. When devtools::check() runs, it fails with a warning that seems to relate to an error in the vignette (shown below). The error cannot be reproduced by running devtools::build_vignettes() which completes without any issue, and indeed every other way of knitting the markdown file seems to work absolutely fine. If I step through the code, it works fine without issues or warnings. While there may be some weird hidden malfunction in my code, it seems wrong that this cannot be detected by running devtools::build_vignettes() and only by going the whole way through devtools::check() which makes it enormously time-consuming to try and debug.

The relevant package/branch is: https://github.com/NathanSkene/EWCE/tree/issue-devCheckFail

To make the issue more complex, the issue seems to be dependent on the machine on which it's being run. It occurs on my mac (OSX 10.15.5) with R 4.0.2 but does not occur on my Linux workstation with R 3.6.1. When TravisCI attempts to check the package, the error also occurs. I'll try reverting the R version on my mac to 3.6.1 and see if it still happens: will report back.

**Error report from devtools::check():**
`
   ----------- FAILURE REPORT -------------- 
   --- failure: the condition has length > 1 ---
   --- srcref --- 
  : 
   --- package (from environment) --- 
  EWCE
   --- call from context --- 
  withCallingHandlers(expr, warning = function(w) if (inherits(w, 
      classes)) tryInvokeRestart("muffleWarning"))
   --- call from argument --- 
  if (!is.na(ctd[[1]])) {
      make_dendro = TRUE
      cells.in.ctd <- function(ctdIN, cells) {
          if (sum(!cells %in% colnames(ctdIN$specificity) == 0)) {
              return(1)
          }
          else {
              return(0)
          }
      }
      if (length(ctd[[1]]$plotting) > 0) {
          annotLevel = which(unlist(lapply(ctd, FUN = cells.in.ctd, 
              cells = as.character(total_res$CellType))) == 1)
          if (length(annotLevel) == 0) {
              stop("All of the cells within total_res should come from a single annotation layer of the CTD")
          }
      }
      if (length(ctd[[annotLevel]]$plotting) > 0) {
          total_res$CellType = factor(total_res$CellType, levels = ctd[[annotLevel]]$plotting$cell_ordering)
      }
  }
   --- R stacktrace ---
  where 1: withCallingHandlers(expr, warning = function(w) if (inherits(w, 
      classes)) tryInvokeRestart("muffleWarning"))
  where 2: suppressWarnings(if (!is.na(ctd[[1]])) {
      make_dendro = TRUE
      cells.in.ctd <- function(ctdIN, cells) {
          if (sum(!cells %in% colnames(ctdIN$specificity) == 0)) {
              return(1)
          }
          else {
              return(0)
          }
      }
      if (length(ctd[[1]]$plotting) > 0) {
          annotLevel = which(unlist(lapply(ctd, FUN = cells.in.ctd, 
              cells = as.character(total_res$CellType))) == 1)
          if (length(annotLevel) == 0) {
              stop("All of the cells within total_res should come from a single annotation layer of the CTD")
          }
      }
      if (length(ctd[[annotLevel]]$plotting) > 0) {
          total_res$CellType = factor(total_res$CellType, levels = ctd[[annotLevel]]$plotting$cell_ordering)
      }
  })
  where 3: ewce.plot(full_results$results, mtc_method = "BH", ctd = ctd)
  where 4: print(ewce.plot(full_results$results, mtc_method = "BH", ctd = ctd)$withDendro)
  where 5: eval(expr, envir, enclos)
  where 6: eval(expr, envir, enclos)
  where 7: withVisible(eval(expr, envir, enclos))
  where 8: withCallingHandlers(withVisible(eval(expr, envir, enclos)), warning = wHandler, 
      error = eHandler, message = mHandler)
  where 9: handle(ev <- withCallingHandlers(withVisible(eval(expr, envir, 
      enclos)), warning = wHandler, error = eHandler, message = mHandler))
  where 10: timing_fn(handle(ev <- withCallingHandlers(withVisible(eval(expr, 
      envir, enclos)), warning = wHandler, error = eHandler, message = mHandler)))
  where 11: evaluate_call(expr, parsed$src[[i]], envir = envir, enclos = enclos, 
      debug = debug, last = i == length(out), use_try = stop_on_error != 
          2L, keep_warning = keep_warning, keep_message = keep_message, 
      output_handler = output_handler, include_timing = include_timing)
  where 12: evaluate::evaluate(...)
  where 13: evaluate(code, envir = env, new_device = FALSE, keep_warning = !isFALSE(options$warning), 
      keep_message = !isFALSE(options$message), stop_on_error = if (options$error && 
          options$include) 0L else 2L, output_handler = knit_handlers(options$render, 
          options))
  where 14: in_dir(input_dir(), evaluate(code, envir = env, new_device = FALSE, 
      keep_warning = !isFALSE(options$warning), keep_message = !isFALSE(options$message), 
      stop_on_error = if (options$error && options$include) 0L else 2L, 
      output_handler = knit_handlers(options$render, options)))
  where 15: block_exec(params)
  where 16: call_block(x)
  where 17: process_group.block(group)
  where 18: process_group(group)
  where 19: withCallingHandlers(if (tangle) process_tangle(group) else process_group(group), 
      error = function(e) {
          setwd(wd)
          cat(res, sep = "\n", file = output %n% "")
          message("Quitting from lines ", paste(current_lines(i), 
              collapse = "-"), " (", knit_concord$get("infile"), 
              ") ")
      })
  where 20: process_file(text, output)
  where 21: knit(input, text = text, envir = envir, quiet = quiet)
  where 22: knit2html(..., force_v1 = TRUE)
  where 23: (if (grepl("\\.[Rr]md$", file)) knit2html_v1 else if (grepl("\\.[Rr]rst$", 
      file)) knit2pandoc else knit)(file, encoding = encoding, 
      quiet = quiet, envir = globalenv(), ...)
  where 24: engine$weave(file, quiet = quiet, encoding = enc)
  where 25: doTryCatch(return(expr), name, parentenv, handler)
  where 26: tryCatchOne(expr, names, parentenv, handlers[[1L]])
  where 27: tryCatchList(expr, classes, parentenv, handlers)
  where 28: tryCatch({
      engine$weave(file, quiet = quiet, encoding = enc)
      setwd(startdir)
      output <- find_vignette_product(name, by = "weave", engine = engine)
      if (!have.makefile && vignette_is_tex(output)) {
          texi2pdf(file = output, clean = FALSE, quiet = quiet)
          output <- find_vignette_product(name, by = "texi2pdf", 
              engine = engine)
      }
      outputs <- c(outputs, output)
  }, error = function(e) {
      thisOK <<- FALSE
      fails <<- c(fails, file)
      message(gettextf("Error: processing vignette '%s' failed with diagnostics:\n%s", 
          file, conditionMessage(e)))
  })
  where 29: tools:::buildVignettes(dir = "/private/var/folders/1k/7sx5jz3j3nq6533gq13ycq_068z7mp/T/RtmpHAglna/EWCE.Rcheck/vign_test/EWCE", 
      ser_elibs = "/var/folders/1k/7sx5jz3j3nq6533gq13ycq_068z7mp/T//RtmpDC5QMU/file23182881ac17.rds")
  
   --- value of length: 7 type: logical ---
                  annot              mean_exp            median_exp 
                   TRUE                  TRUE                  TRUE 
            specificity    median_specificity specificity_quantiles 
                   TRUE                  TRUE                  TRUE 
               plotting 
                   TRUE 
   --- function from context --- 
  function (expr, ...) 
  {
      handlers <- list(...)
      classes <- names(handlers)
      parentenv <- parent.frame()
      if (length(classes) != length(handlers)) 
          stop("condition handlers must be specified with a condition class")
      .Internal(.addCondHands(classes, handlers, parentenv, NULL, 
          TRUE))
      expr
  }
  <bytecode: 0x7ffe7f0c68c0>
  <environment: namespace:base>
   --- function search by body ---
  Function withCallingHandlers in namespace base has this body.
   ----------- END OF FAILURE REPORT -------------- 
  Quitting from lines 333-334 (EWCE.Rmd) 
  Error: processing vignette 'EWCE.Rmd' failed with diagnostics:
  the condition has length > 1
  --- failed re-building ‘EWCE.Rmd’
  
  SUMMARY: processing the following file failed:
    ‘EWCE.Rmd’
  
  Error: Vignette re-building failed.
  Execution halted

> checking installed package size ... NOTE
    installed size is 47.0Mb
    sub-directories of 1Mb or more:
      data  46.2Mb

0 errors ✓ | 1 warning x | 1 note x`

**SessionInfo from Mac:**

`> sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

Random number generation:
 RNG:     Mersenne-Twister 
 Normal:  Inversion 
 Sample:  Rounding 
 
locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] EWCE_0.99.2    colorout_1.2-2

loaded via a namespace (and not attached):
  [1] fs_1.4.2             xopen_1.0.0          usethis_1.6.1        devtools_2.3.1       bit64_0.9-7.1       
  [6] progress_1.2.2       httr_1.4.2           rprojroot_1.3-2      sctransform_0.2.1    tools_4.0.2         
 [11] backports_1.1.8      R6_2.4.1             DBI_1.1.0            BiocGenerics_0.34.0  colorspace_1.4-1    
 [16] withr_2.2.0          tidyselect_1.1.0     gridExtra_2.3        prettyunits_1.1.1    processx_3.4.3      
 [21] bit_1.1-15.2         curl_4.3             compiler_4.0.2       cli_2.0.2            Biobase_2.48.0      
 [26] xml2_1.3.2           desc_1.2.0           ggdendro_0.1-20      labeling_0.3         scales_1.1.1        
 [31] callr_3.4.3          askpass_1.1          rappdirs_0.3.1       stringr_1.4.0        digest_0.6.25       
 [36] rmarkdown_2.3        pkgconfig_2.0.3      htmltools_0.5.0      sessioninfo_1.1.1    dbplyr_1.4.4        
 [41] limma_3.44.3         readxl_1.3.1         rlang_0.4.7          rstudioapi_0.11      RSQLite_2.2.0       
 [46] farver_2.0.3         generics_0.0.2       dplyr_1.0.0          magrittr_1.5         Matrix_1.2-18       
 [51] Rcpp_1.0.5           munsell_0.5.0        S4Vectors_0.26.1     fansi_0.4.1          lifecycle_0.2.0     
 [56] stringi_1.4.6        yaml_2.2.1           MASS_7.3-51.6        pkgbuild_1.1.0       plyr_1.8.6          
 [61] BiocFileCache_1.12.0 grid_4.0.2           blob_1.2.1           parallel_4.0.2       listenv_0.8.0       
 [66] crayon_1.3.4         lattice_0.20-41      cowplot_1.0.0        hms_0.5.3            knitr_1.29          
 [71] ps_1.3.3             pillar_1.4.6         future.apply_1.6.0   reshape2_1.4.4       codetools_0.2-16    
 [76] biomaRt_2.44.1       stats4_4.0.2         pkgload_1.1.0        XML_3.99-0.5         glue_1.4.1          
 [81] evaluate_0.14        remotes_2.2.0        BiocManager_1.30.10  vctrs_0.3.2          cellranger_1.1.0    
 [86] testthat_2.3.2       rcmdcheck_1.3.3      gtable_0.3.0         openssl_1.4.2        purrr_0.3.4         
 [91] future_1.18.0        assertthat_0.2.1     ggplot2_3.3.2        xfun_0.16            roxygen2_7.1.1      
 [96] tibble_3.0.3         RNOmni_0.7.1         AnnotationDbi_1.50.3 memoise_1.1.0        IRanges_2.22.2      
[101] HGNChelper_0.8.1     globals_0.12.5       ellipsis_0.3.1       BiocStyle_2.16.0   `

**SessionInfo from Linux:**
`> sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 19.04

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.3.5.so

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8       
 [4] LC_COLLATE=en_GB.UTF-8     LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] EWCE_0.99.2    colorout_1.2-1

loaded via a namespace (and not attached):
 [1] fs_1.4.1             xopen_1.0.0          usethis_1.6.1        devtools_2.3.0      
 [5] bit64_0.9-7          progress_1.2.2       httr_1.4.1           rprojroot_1.3-2     
 [9] sctransform_0.2.1    tools_3.6.1          backports_1.1.8      R6_2.4.1            
[13] DBI_1.1.0            BiocGenerics_0.32.0  colorspace_1.4-1     withr_2.2.0         
[17] tidyselect_1.1.0     gridExtra_2.3        prettyunits_1.1.1    processx_3.4.3      
[21] bit_1.1-15.2         curl_4.3             compiler_3.6.1       cli_2.0.2           
[25] Biobase_2.46.0       xml2_1.3.2           desc_1.2.0           ggdendro_0.1-20     
[29] scales_1.1.1         callr_3.4.3          askpass_1.1          rappdirs_0.3.1      
[33] stringr_1.4.0        digest_0.6.25        pkgconfig_2.0.3      sessioninfo_1.1.1   
[37] dbplyr_1.4.4         limma_3.42.2         rlang_0.4.7          readxl_1.3.1        
[41] rstudioapi_0.11      RSQLite_2.2.0        dplyr_0.8.5          magrittr_1.5        
[45] Matrix_1.2-18        Rcpp_1.0.5           munsell_0.5.0        S4Vectors_0.24.4    
[49] fansi_0.4.1          lifecycle_0.2.0      stringi_1.4.6        MASS_7.3-51.4       
[53] pkgbuild_1.1.0       plyr_1.8.6           BiocFileCache_1.10.2 grid_3.6.1          
[57] blob_1.2.1           parallel_3.6.1       listenv_0.8.0        crayon_1.3.4        
[61] lattice_0.20-38      cowplot_1.0.0        hms_0.5.3            knitr_1.28          
[65] ps_1.3.3             pillar_1.4.6         future.apply_1.5.0   reshape2_1.4.4      
[69] codetools_0.2-16     biomaRt_2.42.1       stats4_3.6.1         pkgload_1.1.0       
[73] XML_3.99-0.3         glue_1.4.1           packrat_0.5.0        remotes_2.1.1       
[77] vctrs_0.3.2          testthat_2.3.2       cellranger_1.1.0     gtable_0.3.0        
[81] openssl_1.4.1        purrr_0.3.4          rcmdcheck_1.3.3      future_1.17.0       
[85] assertthat_0.2.1     ggplot2_3.3.2        xfun_0.14            roxygen2_7.1.0      
[89] tibble_3.0.3         RNOmni_0.7.1         AnnotationDbi_1.48.0 memoise_1.1.0       
[93] IRanges_2.20.2       HGNChelper_0.8.1     globals_0.12.5       ellipsis_0.3.1`

**Comment by NathanSkene** [comment]
I've now tried running devtools::check() on my mac within a docker image setup with R 3.6. As I suspected, it works fine and the check() finishes without any issues. So this suggests that the error is not with my code but with either devtools or something else related to the R 4.0 upgrade.

The dockerfile I used is here: https://github.com/NathanSkene/EWCE/blob/issue-devCheckFail/Dockerfile

**Comment by hadley** [comment]
In our experience this is almost always a subtle and intensely frustrating bug. We are collecting hints for debugging this sort of thing at https://github.com/hadley/r-pkgs/issues/483 but generally your best bet is to work to create a reprex as minimal as possible so eventually the precise problem becomes known and you can figure out what makes the check environment different.

**Comment by NathanSkene** [comment]
Figured out the issue. Did what you suggested and trimmed this down to a minimally reproducible example (https://github.com/NathanSkene/EWCE/commits/dev-VignetteCheckMinimal). Realised that the issue is that I used a suppressWarnings call in the offending function. This meant that a warning that gets issued is ignored and not reported by devtools::build_vignette() but is still noticed by devtools::check() and results in a failure.

Seems like a more consistent treatment of warnings from within suppressWarnings() would be useful between devtools::check() and devtools::build_vignette(). Otherwise other folks are likely to get stuck on this rock as well. I acknowledge it's probably bad practise to use supresssWarnings within a function... but if this is verboten then it should result in it's own error within devtools::check(). 

As noted above, it only seems to be in R > 4.0 that this has begun to occur. 

Anyway, thanks for your help (and for developing all these great tools in the first place!).
